corehq/form_processor/submission_post.py

```diff
@@ -500,14 +497,30 @@ def index_case_search(instance, case_models):
 
     @staticmethod
     @tracer.wrap(name='submission.process_cases_and_stock')
-    def process_xforms_for_cases(xforms, case_db, timing_context=None):
+    def process_xforms_for_cases(xforms, case_db, domain=None, timing_context=None):
         from casexml.apps.case.xform import process_cases_with_casedb
         from corehq.apps.commtrack.processing import process_stock
 
-        timing_context = timing_context or TimingContext()
-
         instance = xforms[0]
 
+        # INSERT "NOT" HERE when done w testing
+        if ALLOW_SUBMISSION_WITHOUT_METADATA.enabled(domain):
+            form_data = instance.form_data
+            if 'case' in form_data:
+                if 'create' in form_data['case']:
+                    for tag in required_create_tags:
+                        if tag not in form_data['case']['create']:
+                            raise MetadataError('One or more required tags missing in the create node')
+                elif 'update' in form_data['case'] and not isinstance(form_data['case']['update'], dict):
+                    raise MetadataError('Missing nodes in the update node.')
```

Comment:
This is only handling a case in the root of the data node but cases can appear anywhere in the XML.

I would recommend using the existing tools for extracting case blocks [casexml.apps.case.xform.extract_case_blocks](https://github.com/dimagi/commcare-hq/blob/3b2d4586b18ff87f5b1eed26d28350a9392d5cf8/corehq/ex-submodules/casexml/apps/case/xform.py#L181-L202) or else pushing this down into the case processing layer e.g. adding validation to [CaseCreateAction](https://github.com/dimagi/commcare-hq/blob/cba51a39deebb72248c12c1597ec1b9951cefd8a/corehq/ex-submodules/casexml/apps/case/xml/parser.py#L139-L140)

Reply:
thanks for pointing this out - I had no idea something like this existed!